### PR TITLE
autocomplete tags/components

### LINKED ISSUE #244: ✨ Autocomplete angle bracketed components the same way braces are autocompleted
### Description

When a user uses angle brackets with a component, for instance typing `<div>`, and then presses enter, the system autocompletes that component, adding `</div>` on the line below the new current line and indents the new current line.

### Alternatives Considered

_No response_

### Additional Context

_No response_

### Screenshots

Screenshot 2024-04-27 at 2 10 06 PM

### PR DESCRIPTION
**Pull request by knotbin**
### Description

Tags in HTML, JS, TS, JSX, and TSX are now autocompleted. When you type `<div>` for example, the closing tag `</div>` will be autocompleted. If you press enter, you will be put on a new line in between the opening and closing and that new line will be indented. All tag attributes are ignored in the closing tag.

### Related Issues

* closes #244

### Checklist

- [x] I read and understood the [contributing guide](https://github.com/CodeEditApp/CodeEdit/blob/main/CONTRIBUTING.md) as well as the [code of conduct](https://github.com/CodeEditApp/CodeEdit/blob/main/CODE_OF_CONDUCT.md)
- [x] The issues this PR addresses are related to each other
- [x] My changes generate no new warnings
- [x] My code builds and runs on my machine
- [x] My changes are all related to the related issue above
- [x] I documented my code

### Screenshots

https://github.com/CodeEditApp/CodeEditSourceEditor/assets/118622417/cf9ffe27-7592-49d5-bee8-edacdd6ab5f4

### CODE REVIEW

**Comment by austincondiff** [comment]
Too inconsistent with other auto completed brackets, The completion should occur when you type the next character after >. For example…

1. <|
2. <d|
3. <di|
4. <div|
5. <div>|
6 <div>X|</div> (autocomplete occurred after typing X)

**Comment by austincondiff** [comment]
When I said that in the issue discussion, I forgot about how Xcode autocompletes brackets, my mistake. We should probably be consistent with how Xcode handles autocompleted pairs by defaulting to after the first character typed including return as mentioned above. Again, this will eventually be configurable via settings.

**Comment by knotbin** [comment]
I think this is different from pairs as many times you want to put nothing in the middle and it would be tedious to have to hit space and backspace or type out the entire closing tag, unlike with pairs which are only one character long.

**Comment by austincondiff** [comment]
I can see where you are coming from. I appreciate your patience in this discussion and I hope you don't mind the back and forth. Just a few points to consider...

> using self closing tags is frowned upon in HTML
While true as of recent, it is still pretty common to do this in JSX.

> I think this is different from pairs as many times you want to put nothing in the middle [...] Making the user type space and backspace would make the development experience much less smooth and a closing tag is much longer to type out than a single closing bracket.
This shouldn't be necessary to get this desired outcome. Just as you can type `]` after a `[' to close an array for example, you can type `</` after completing an opening tag to complete the closing tag and place the cursor at the end.

This does make me consider though, what if the desired line was
```
<div><p>Hello world!</p></div>
```

1. `<div><p>Hello world!</p></div>`
2. `<div><p>Hello world!</p></div>`

**Comment by knotbin** [comment]
This last case you presented is why I believe it's much better to have it work as it currently does. Nesting components is much easier when the closing tag is automatically placed and all of that extra logic you described isn't needed and the dev can just add the nested component without any weird autofill confusion. I feel like people wouldn't know that a closing tag will autofill if it is not appearing, and might just go to type it themselves instead of the nested div. Overall, this feels overly complicated and confusing for something that I believe has the best and most simple solution already implemented. 

I think the nesting tabs use case is the biggest argument for the way it currently works, as it is much less confusing for the dev to nest components when they can see the closed component they are nesting in.

The use case of a tag with nothing in between open and close also still feels bad to me, as if the user has to type </, the purpose of the autofill is kind of defeated. I think in this case, it would be very disorienting for webdevs for it to require a space or extra typing to autofill. I think that there is a very good case that tags work very differently from brackets and therefore shouldn't have consistent logic with them. 
It feels like the obvious solution is to have it complete immediately to minimize confusion and extra typing, and make nesting tabs more fluid.

**Comment by austincondiff** [comment]
It might be interesting to also explore attributes/props. This might look like...

1. `<p|`
2. `<p |></p>` (space was pressed - we know intent is to add attributes so we can autocomplete tag but keep the cursor where it is so the user can enter attributes)
3. `<p class="|></p>`
4. `<p class="l|"></p>`
5. `<p class="large|"></p>`

**Comment by austincondiff** [comment]
I agree that your current approach is definitely simpler and to be fair, a PR is not the place to explore this. Let's go ahead and merge this and this can be discussed at a later point in time.

File: Sources/CodeEditSourceEditor/Controller/TextViewController+TextFormation.swift [hunk 1]
```diff
@@ -40,6 +40,7 @@ extension TextViewController {
         setUpNewlineTabFilters(indentOption: indentOption)
         setUpDeletePairFilters(pairs: BracketPairs.allValues)
         setUpDeleteWhitespaceFilter(indentOption: indentOption)
+        setUpTagFiler()
```
Comment (austincondiff) [inline Sources/CodeEditSourceEditor/Controller/TextViewController+TextFormation.swift:43]: Small typo here...
```suggestion
        setUpTagFilter()
```

File: Sources/CodeEditSourceEditor/Controller/TextViewController+TextFormation.swift [hunk 2]
```diff
@@ -90,6 +91,18 @@ extension TextViewController {
         textFilters.append(filter)
     }
 
+    private func setUpTagFiler() {
```
Comment (austincondiff) [inline Sources/CodeEditSourceEditor/Controller/TextViewController+TextFormation.swift:94]: Small typo here...
```suggestion
    private func setUpTagFilter() {
```

**Review by austincondiff — CHANGES_REQUESTED** [review]
Looks good, just a few small things

**Comment by knotbin** [comment]
Sorry, should have caught that 😅

**Review by austincondiff — APPROVED** [review]

**Pull request merged**